RFC: The `kani::should_panic` attribute

adpaco-aws opened the pull request:
### Description of changes: 

Users may want to express that a verification harness should panic.
This RFC proposes a new harness attribute `#[kani::should_panic]` that informs Kani about this expectation.

Rendered preview [here](https://github.com/adpaco-aws/rmc/blob/rfc-should-fail/rfc/src/rfcs/0005-should-fail-attr.md).

### Resolved issues:

Related #600 

### Call-outs:

<!-- 
Address any potentially confusing code. Is there code added that needs to be cleaned up later? Is there code that is missing because it’s still in development? 
-->

### Testing:

* How is this change tested? N/A

* Is this a refactor change? N/A

### Checklist
- [x] Each commit message has a non-empty body, explaining why the change was made
- [ ] Methods or procedures are documented
- [ ] Regression or unit tests are included, or existing tests cover the modified code
- [x] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

celinval commented on rfc/src/rfcs/0005-should-fail-attr.md:
Doesn't this conflict with the detailed representation?

adpaco-aws replied:
No, this shows up for both the default and terse outputs.

celinval commented on rfc/src/rfcs/0005-should-fail-attr.md:
I'm so glad you are tackling this here, but I wonder if we can still improve this. Double negation is hard to parse.

adpaco-aws replied:
True, it's also not clear to me how this result will look in #1690. I need to spend more time on this one.

adpaco-aws replied:
Couldn't find a simpler representation. It's now a question in the "Open questions", and I'll explicitly look for feedback on this.

celinval commented on rfc/src/rfcs/0005-should-fail-attr.md:
+1.

celinval commented on rfc/src/rfcs/0005-should-fail-attr.md:
Did you consider adding `#[should_panic]` instead of `#[should_fail]` for now. I am trying to see what is the motivation for users to knowingly leave code that triggers UB.

celinval commented on rfc/src/rfcs/0005-should-fail-attr.md:
could users have multiple annotations?

adpaco-aws replied:
Another good point, I'll add that to this list.

adpaco-aws replied:
The RFC now promotes a discussion around multiple panics, and suggests the panic expectation be limited to one. It isn't clear to me that we want to allow multiple panics with this attribute, but there's an open question related to this aspect.

celinval commented on rfc/src/rfcs/0005-should-fail-attr.md:
I'm worried that `should_fail` is too open since it includes any UB.

adpaco-aws replied:
This was answered in my long comment.

karkhaz commented on rfc/src/rfcs/0005-should-fail-attr.md:
This seems like a big deal TBH. From basically the first example, I was wondering "what happens if a new unintentional `panic` later gets inserted earlier in the code?"

Is there a way to 'fix' this? Instead of an attribute, could we have a magic comment (or some Rusty equivalent) directly above the line that is expected to panic, instead of the whole harness?

adpaco-aws replied:
The main problem with the approach you mention (and similar ones like the "function that takes a closure", which is now mentioned as an alternative) is that you can't use them on code that's external to your packages.

Moreover, it seems to me that users just want a coarse mechanism to check that a panic is reachable. It's possible that users want a more fine-grained mechanism, as noted in the RFC, but we simply can't get much data on that unless we have the coarse one first.

robdockins commented on rfc/src/rfcs/0005-should-fail-attr.md:
A third potential case is that the user has used `cfg` attributes to stub out part of the code, and wants to be sure the expected panic occurs both in the code as exposed to Kani and in the "normal" code.

In the case I'm concerned with, I used stubs to implement the low-level concurrency primitives that are the basis of the library, so I was wanting to be sure the code with stubs also behaved as expected.

adpaco-aws replied:
Thanks @robdockins , added the scenario you describe as a third reason.

adpaco-aws commented:
Addressed comments on the PR and from our discussion yesterday:
 * Moved discussion about multiple panics to UX section, and extended it significantly.
 * Provide answers for all questions that were open, and move them to a subsection "resolved questions".
 * Add new alternative to mention the Kani-API-that-takes-closure approach.
 * Modified representations, and selected the second one as the recommended.
 * Added third motivating reason for the attribute.

zhassan-aws commented on rfc/src/rfcs/0005-should-fail-attr.md:
````diff
@@ -88,24 +88,24 @@ Since we added `#[kani::should_panic]`, running this example would produce a suc
 
 ```rust
-VERIFICATION:- SUCCESSFUL (expected FAILED and it FAILED)
+VERIFICATION:- SUCCESSFUL (expected and verified one panic at least)
````
I would avoid associating "verified" with "panic". Maybe say `encountered one or more panics as expected`?

adpaco-aws replied:
Thanks! I've changed it and extended this section with data from the actual implementation (see #2315)

zhassan-aws reviewed (approved):
Looks good. Just one comment.
